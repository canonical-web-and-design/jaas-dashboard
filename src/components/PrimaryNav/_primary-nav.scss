@import "../../scss/settings";
@import "../../scss/functions/z-index";
@import "vanilla-framework/scss/vanilla";

.p-primary-nav {
  display: flex;
  flex-direction: column;
  height: 100%;
  position: relative;
  width: $sidebar-full-width;

  a,
  a:hover {
    color: $color-light;
    text-decoration: none;
  }

  &__logo {
    display: flex;
    margin-bottom: 0.5rem;
    margin-left: 2px;
    padding-top: 3px;

    &-text {
      padding-left: 0.6rem;
    }
  }

  &__header {
    background: $color-navigation-dark;
    display: grid;
    grid-template-columns: 1fr 40px;
    padding: 0.5rem 0.5rem 0;
    position: relative;
  }

  &__toggle {
    align-items: center;
    background: none;
    border: 0;
    display: grid;
    margin: 0;
    padding: 0;

    .p-icon--contextual-menu {
      margin: auto;
    }

    &:hover,
    &:active {
      background: none !important;
    }

    &:focus {
      outline: 1px dotted $color-navigation-highlight;
    }
  }

  .p-list__item {
    padding: 0;

    .p-list__link {
      border-left: 3px solid transparent;
      color: $color-light;
      display: flex;
      padding: 0.25rem 0;
      padding-left: calc(1rem - 3px);
      width: 100%;

      &:hover {
        background-color: $color-navigation-highlight;
      }

      &.is-selected {
        background-color: $color-navigation-highlight;
        border-left-color: $color-navigation-accent;
      }
    }

    .p-list__icon {
      height: 16px;
      padding-right: 1.25rem;
      position: relative;
      top: 0.275rem;
    }
  }

  .p-primary-nav__bottom {
    border-top: 1px solid #666;
    margin-top: 0;
    padding-top: 1rem;

    .p-list {
      &__item {
        color: #fff;
        display: flex;

        .p-label--new {
          margin-left: auto;
          margin-right: 1rem;
          margin-top: 0.5rem;
        }

        .version {
          color: $color-light;
          padding-left: 3.25rem;
          padding-top: 0.5rem;
        }
      }
    }
  }

  .p-list.is-external {
    @include vf-animation(max-height, brisk, ease-in-out);

    background: $color-navigation-dark;
    margin-bottom: 0;
    max-height: 0;
    overflow: hidden;

    .p-list__link {
      padding-left: calc(3.25rem);
    }
  }

  .p-primary-nav__toggle {
    position: relative;
    top: -3px;
  }

  .l-side:not(.is-collapsed) &.ext-nav-open {
    .is-external {
      max-height: 200px;
      padding-bottom: 0.5rem;
    }

    .p-primary-nav__toggle {
      position: relative;
      top: -3px;
      transform: rotate(180deg);
    }
  }

  .entity-count {
    background-color: $color-negative;
    border-radius: 50%;
    display: inline-block;
    font-size: 0.875rem;
    height: 1.5rem;
    margin-left: auto;
    margin-right: 1rem;
    padding: 0 0.25rem;
    text-align: center;
    width: 1.5rem;
  }
}
