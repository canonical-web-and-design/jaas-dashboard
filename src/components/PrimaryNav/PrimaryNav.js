import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { NavLink, useLocation } from "react-router-dom";
import classNames from "classnames";

import { getGroupedModelStatusCounts } from "app/selectors";

import { externalNavActive } from "ui/actions";
import { isExternalNavActive } from "ui/selectors";

import UserMenu from "components/UserMenu/UserMenu";

// Image imports
import logoMark from "static/images/logo/logo-mark.svg";
import logoText from "static/images/logo/logo-text.svg";
import modelsIcon from "static/images/icons/models-icon.svg";
import modelsIconSelected from "static/images/icons/models-icon--selected.svg";
import controllersIcon from "static/images/icons/controllers-icon.svg";
import controllersIconSelected from "static/images/icons/controllers-icon--selected.svg";
// Remove these nav links until these sections are active
// import usageIcon from "static/images/icons/usage-icon.svg";
// import usageIconSelected from "static/images/icons/usage-icon--selected.svg";
// import logsIcon from "static/images/icons/logs-icon.svg";
// import logsIconSelected from "static/images/icons/logs-icon--selected.svg";

// Style imports
import "./_primary-nav.scss";

const pages = [
  {
    label: "Models",
    path: "/models",
    icon: modelsIcon,
    iconSelected: modelsIconSelected,
  },
  {
    label: "Controllers",
    path: "/controllers",
    icon: controllersIcon,
    iconSelected: controllersIconSelected,
  },
  // Remove these nav links until these sections are active
  // { label: "Usage", path: "/usage", icon: usageIcon, iconSelected: usageIconSelected },
  // { label: "Logs", path: "/logs", icon: logsIcon, iconSelected: logsIconSelected },
];

const PrimaryNav = () => {
  const [activeLinkValue, setActiveLinkValue] = useState("");
  const extNavActive = useSelector(isExternalNavActive);
  const { blocked } = useSelector(getGroupedModelStatusCounts);
  const location = useLocation();
  const dispatch = useDispatch();

  useEffect(() => {
    setActiveLinkValue(location.pathname);
  }, [location.pathname]);

  useEffect(() => {
    // Inject the the Usabilla script.
    // prettier-ignore
    /* eslint-disable */
    window.lightningjs || function (c) { function g(b, d) { d && (d += (/\?/.test(d) ? "&" : "?") + "lv=1"); c[b] || function () { var i = window, h = document, j = b, g = h.location.protocol, l = "load", k = 0; (function () { function b() { a.P(l); a.w = 1; c[j]("_load") } c[j] = function () { function m() { m.id = e; return c[j].apply(m, arguments) } var b, e = ++k; b = this && this != i ? this.id || 0 : 0; (a.s = a.s || []).push([e, b, arguments]); m.then = function (b, c, h) { var d = a.fh[e] = a.fh[e] || [], j = a.eh[e] = a.eh[e] || [], f = a.ph[e] = a.ph[e] || []; b && d.push(b); c && j.push(c); h && f.push(h); return m }; return m }; var a = c[j]._ = {}; a.fh = {}; a.eh = {}; a.ph = {}; a.l = d ? d.replace(/^\/\//, (g == "https:" ? g : "http:") + "//") : d; a.p = { 0: +new Date }; a.P = function (b) { a.p[b] = new Date - a.p[0] }; a.w && b(); i.addEventListener ? i.addEventListener(l, b, !1) : i.attachEvent("on" + l, b); var q = function () { function b() { return ["<head></head><", c, ' onload="var d=', n, ";d.getElementsByTagName('head')[0].", d, "(d.", g, "('script')).", i, "='", a.l, "'\"></", c, ">"].join("") } var c = "body", e = h[c]; if (!e) return setTimeout(q, 100); a.P(1); var d = "appendChild", g = "createElement", i = "src", k = h[g]("div"), l = k[d](h[g]("div")), f = h[g]("iframe"), n = "document", p; k.style.display = "none"; e.insertBefore(k, e.firstChild).id = o + "-" + j; f.frameBorder = "0"; f.id = o + "-frame-" + j; /MSIE[ ]+6/.test(navigator.userAgent) && (f[i] = "javascript:false"); f.allowTransparency = "true"; l[d](f); try { f.contentWindow[n].open() } catch (s) { a.domain = h.domain, p = "javascript:var d=" + n + ".open();d.domain='" + h.domain + "';", f[i] = p + "void(0);" } try { var r = f.contentWindow[n]; r.write(b()); r.close() } catch (t) { f[i] = p + 'd.write("' + b().replace(/"/g, String.fromCharCode(92) + '"') + '");d.close();' } a.P(2) }; a.l && setTimeout(q, 0) })() }(); c[b].lv = "1"; return c[b] } var o = "lightningjs", k = window[o] = g(o); k.require = g; k.modules = c }({});
    window.usabilla_live = lightningjs.require(
      "usabilla_live",
      "//w.usabilla.com/a0ba0c2df8fd.js"
    );
    // Hide Usabilla default button
    window.usabilla_live("hide");
    /* eslint-enable */
  }, []);

  return (
    <nav
      className={classNames("p-primary-nav", {
        "ext-nav-open": extNavActive,
      })}
    >
      <div className="p-primary-nav__header">
        <a href="https://jaas.ai" className="p-primary-nav__logo">
          <img
            className="p-primary-nav__logo-icon"
            src={logoMark}
            alt="JAAS logo"
            height="30"
            width="30"
          />
          <img
            className="p-primary-nav__logo-text"
            src={logoText}
            height="30"
            alt=""
          />
        </a>
        <button
          className="p-primary-nav__toggle"
          onClick={() => dispatch(externalNavActive(!extNavActive))}
        >
          <i className="p-icon--contextual-menu">Toggle external navigation</i>
        </button>
      </div>
      <ul className="p-list is-external">
        <li className="p-list__item">
          <a
            className="p-list__link"
            href="https://jaas.ai/store"
            target="_blank"
            rel="noopener noreferrer"
          >
            Store
          </a>
        </li>
        <li className="p-list__item">
          <a
            className="p-list__link"
            href="https://jaas.ai/jaas"
            target="_blank"
            rel="noopener noreferrer"
          >
            About
          </a>
        </li>
        <li className="p-list__item">
          <a
            className="p-list__link"
            href="https://jaas.ai/how-it-works"
            target="_blank"
            rel="noopener noreferrer"
          >
            How it works
          </a>
        </li>
        <li className="p-list__item">
          <a
            className="p-link--external p-list__link"
            href="https://discourse.jujucharms.com/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Discourse
          </a>
        </li>
        <li className="p-list__item">
          <a
            className="p-link--external p-list__link"
            href="https://jaas.ai/docs/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Docs
          </a>
        </li>
      </ul>
      <ul className="p-list is-internal">
        {pages.map((navItem) => (
          <li key={navItem.path} className="p-list__item">
            <NavLink
              className="p-list__link"
              isActive={(match) => {
                if (match && match.url.includes(navItem.path)) {
                  return true;
                }
              }}
              to={navItem.path}
              activeClassName="is-selected"
            >
              <img
                className="p-list__icon"
                src={
                  activeLinkValue === navItem.path
                    ? navItem.iconSelected
                    : navItem.icon
                }
                alt={`${navItem.label} icon`}
              />
              {navItem.label}
              {navItem.label === "Models" && blocked > 0 ? (
                <span className="entity-count">{blocked}</span>
              ) : (
                ""
              )}
            </NavLink>
          </li>
        ))}
      </ul>
      <div className="p-primary-nav__bottom">
        <ul className="p-list">
          <li className="p-list__item">
            <a
              className="p-list__link"
              href="https://github.com/canonical-web-and-design/jaas-dashboard/issues/new"
              target="_blank"
              rel="noopener noreferrer"
            >
              Report a bug
            </a>
          </li>
          <li className="p-list__item">
            <a
              rel="noopener noreferrer"
              target="_blank"
              className="p-list__link"
              href="#_"
              onClick={(e) => {
                e.preventDefault();
                window.usabilla_live("click");
              }}
            >
              Give feedback
            </a>
          </li>
          <li className="p-list__item">
            <span className="version">Version 0.0.2</span>
            <span className="p-label--new">Alpha</span>
          </li>
        </ul>
      </div>
      <UserMenu />
    </nav>
  );
};

export default PrimaryNav;
