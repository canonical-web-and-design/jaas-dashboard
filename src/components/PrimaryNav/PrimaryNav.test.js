import React from "react";
import { BrowserRouter as Router, MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import dataDump from "testing/complete-redux-store-dump";

import PrimaryNav from "./PrimaryNav";

const mockStore = configureStore([]);
describe("Primary Nav", () => {
  it("toggles external nav menu", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <PrimaryNav />
        </Router>
      </Provider>
    );

    const primaryNav = ".p-primary-nav";
    const primaryNavToggle = ".p-primary-nav__toggle";
    expect(wrapper.find(primaryNav).hasClass("ext-nav-open")).toEqual(false);
    wrapper.find(primaryNavToggle).simulate("click");
    const actions = store.getActions();
    const expectedPayload = { payload: true, type: "TOGGLE_EXTERNAL_NAV" };
    expect(actions).toEqual([expectedPayload]);
  });

  it("applies is-selected state correctly", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/controllers"]}>
          <PrimaryNav />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("a.is-selected").text()).toStrictEqual("Controllers");
  });

  it("displays correct number of blocked models", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/"]}>
          <PrimaryNav />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".entity-count").text()).toStrictEqual("6");
  });
});
