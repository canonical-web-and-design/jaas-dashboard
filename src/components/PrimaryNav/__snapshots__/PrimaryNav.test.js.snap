// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary Nav renders without crashing and matches snapshot 1`] = `
<nav
  className="p-primary-nav"
>
  <div
    className="p-primary-nav__header"
  >
    <Link
      className="p-primary-nav__logo"
      to="/"
    >
      <LinkAnchor
        className="p-primary-nav__logo"
        href="/"
        navigate={[Function]}
      >
        <a
          className="p-primary-nav__logo"
          href="/"
          onClick={[Function]}
        >
          <img
            alt="JAAS logo"
            className="p-primary-nav__logo-icon"
            height="30"
            src="logo-mark.svg"
            width="30"
          />
          <img
            alt=""
            className="p-primary-nav__logo-text"
            height="30"
            src="logo-text.svg"
          />
        </a>
      </LinkAnchor>
    </Link>
    <button
      className="p-primary-nav__toggle"
      onClick={[Function]}
    >
      <i
        className="p-icon--contextual-menu"
      >
        Toggle external navigation
      </i>
    </button>
  </div>
  <ul
    className="p-list is-external"
  >
    <li
      className="p-list__item"
    >
      <a
        className="p-list__link"
        href="https://jaas.ai/store"
        rel="noopener noreferrer"
        target="_blank"
      >
        Store
      </a>
    </li>
    <li
      className="p-list__item"
    >
      <a
        className="p-list__link"
        href="https://jaas.ai/about"
        rel="noopener noreferrer"
        target="_blank"
      >
        About
      </a>
    </li>
    <li
      className="p-list__item"
    >
      <a
        className="p-list__link"
        href="https://jaas.ai/how-it-works"
        rel="noopener noreferrer"
        target="_blank"
      >
        How it works
      </a>
    </li>
    <li
      className="p-list__item"
    >
      <a
        className="p-link--external p-list__link"
        href="https://discourse.jujucharms.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Discourse
      </a>
    </li>
    <li
      className="p-list__item"
    >
      <a
        className="p-link--external p-list__link"
        href="https://jaas.ai/docs/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Docs
      </a>
    </li>
  </ul>
  <ul
    className="p-list is-internal"
  >
    <li
      className="p-list__item"
      key="/models"
    >
      <NavLink
        activeClassName="is-selected"
        className="p-list__link"
        isActive={[Function]}
        to="/models"
      >
        <Link
          aria-current={null}
          className="p-list__link"
          to={
            Object {
              "hash": "",
              "pathname": "/models",
              "search": "",
              "state": null,
            }
          }
        >
          <LinkAnchor
            aria-current={null}
            className="p-list__link"
            href="/models"
            navigate={[Function]}
          >
            <a
              aria-current={null}
              className="p-list__link"
              href="/models"
              onClick={[Function]}
            >
              <img
                alt="Models icon"
                className="p-list__icon"
                src="models-icon.svg"
              />
              Models
              <span
                className="entity-count"
              >
                2
              </span>
            </a>
          </LinkAnchor>
        </Link>
      </NavLink>
    </li>
    <li
      className="p-list__item"
      key="/controllers"
    >
      <NavLink
        activeClassName="is-selected"
        className="p-list__link"
        isActive={[Function]}
        to="/controllers"
      >
        <Link
          aria-current={null}
          className="p-list__link"
          to={
            Object {
              "hash": "",
              "pathname": "/controllers",
              "search": "",
              "state": null,
            }
          }
        >
          <LinkAnchor
            aria-current={null}
            className="p-list__link"
            href="/controllers"
            navigate={[Function]}
          >
            <a
              aria-current={null}
              className="p-list__link"
              href="/controllers"
              onClick={[Function]}
            >
              <img
                alt="Controllers icon"
                className="p-list__icon"
                src="controllers-icon.svg"
              />
              Controllers
            </a>
          </LinkAnchor>
        </Link>
      </NavLink>
    </li>
    <li
      className="p-list__item"
      key="/usage"
    >
      <NavLink
        activeClassName="is-selected"
        className="p-list__link"
        isActive={[Function]}
        to="/usage"
      >
        <Link
          aria-current={null}
          className="p-list__link"
          to={
            Object {
              "hash": "",
              "pathname": "/usage",
              "search": "",
              "state": null,
            }
          }
        >
          <LinkAnchor
            aria-current={null}
            className="p-list__link"
            href="/usage"
            navigate={[Function]}
          >
            <a
              aria-current={null}
              className="p-list__link"
              href="/usage"
              onClick={[Function]}
            >
              <img
                alt="Usage icon"
                className="p-list__icon"
                src="usage-icon.svg"
              />
              Usage
            </a>
          </LinkAnchor>
        </Link>
      </NavLink>
    </li>
    <li
      className="p-list__item"
      key="/logs"
    >
      <NavLink
        activeClassName="is-selected"
        className="p-list__link"
        isActive={[Function]}
        to="/logs"
      >
        <Link
          aria-current={null}
          className="p-list__link"
          to={
            Object {
              "hash": "",
              "pathname": "/logs",
              "search": "",
              "state": null,
            }
          }
        >
          <LinkAnchor
            aria-current={null}
            className="p-list__link"
            href="/logs"
            navigate={[Function]}
          >
            <a
              aria-current={null}
              className="p-list__link"
              href="/logs"
              onClick={[Function]}
            >
              <img
                alt="Logs icon"
                className="p-list__icon"
                src="logs-icon.svg"
              />
              Logs
            </a>
          </LinkAnchor>
        </Link>
      </NavLink>
    </li>
  </ul>
  <div
    className="p-primary-nav__bottom"
  >
    <ul
      className="p-list"
    >
      <li
        className="p-list__item"
      >
        <a
          className="p-list__link"
          href="https://github.com/canonical-web-and-design/jaas-dashboard/issues/new"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="bug icon"
            className="p-list__icon"
            src="bug-icon.svg"
          />
          Report a bug
        </a>
      </li>
    </ul>
  </div>
</nav>
`;
